src/xb-builder-source.c

```diff
@@ -394,31 +453,77 @@ xb_builder_source_get_info (XbBuilderSource *self)
 	return priv->info;
 }
 
+/* converts foo.xml.gz to foo.xml */
+static void
+xb_builder_source_remove_last_extension (gchar *basename)
+{
+	gchar *tmp = g_strrstr (basename, ".");
+	if (tmp != NULL)
+		*tmp = '\0';
+}
+
 GInputStream *
 xb_builder_source_get_istream (XbBuilderSource *self,
 			       GCancellable *cancellable,
 			       GError **error)
 {
-	XbBuilderSourceConverterItem *item;
 	XbBuilderSourcePrivate *priv = GET_PRIVATE (self);
+	XbBuilderSourceAdapter *item;
+	g_autofree gchar *basename = NULL;
 
 	g_return_val_if_fail (XB_IS_BUILDER_SOURCE (self), NULL);
 
 	/* nothing required */
 	if (priv->istream != NULL)
 		return g_object_ref (priv->istream);
 
-	/* decompress if required */
+	/* decompress if required using the deprecated vfunc */
 	item = xb_builder_source_get_converter_by_mime (self, priv->content_type);
-	if (item == NULL) {
-		g_set_error (error,
-			     G_IO_ERROR,
-			     G_IO_ERROR_NOT_SUPPORTED,
-			     "cannot process file of type %s",
-			     priv->content_type);
-		return NULL;
+	if (item != NULL && item->func_file != NULL) {
+		return item->func_file (self, priv->file, item->user_data,
+					cancellable, error);
 	}
-	return item->func (self, priv->file, item->user_data, cancellable, error);
+
+	/* convert the file to a GFileInputStream */
+	priv->istream = G_INPUT_STREAM (g_file_read (priv->file, cancellable, error));
+	if (priv->istream == NULL)
+		return NULL;
+
+	/* run the content type handlers until we get application/xml */
+	basename = g_file_get_basename (priv->file);
+	do {
+		XbBuilderSourceCtx *ctx = xb_builder_source_ctx_new (priv->istream);
+		g_autofree gchar *content_type = NULL;
+		g_autoptr(GInputStream) istream_tmp = NULL;
+
+		/* get the content type of the stream */
+		xb_builder_source_ctx_set_filename (ctx, basename);
+		content_type = xb_builder_source_ctx_get_content_type (ctx,
+								       cancellable,
+								       error);
+		if (content_type == NULL)
+			return NULL;
+		if (g_strcmp0 (content_type, "application/xml") == 0)
```

Comment:
Considering #9, is looking for text/plain another alternative?  Or should #9 be rejected outright and don't support unless proper mime types are used?

Reply:
`text/plain` isn't right. I think #9 should be closed -- I'd be fine adding some heuristics to `xb_builder_source_ctx_get_content_type()` if content-type detection can't be enabled for some reason.